Changed World Generation to use random seed numbers if not specified

TheShermanTanker opened the pull request:
Currently, Cloudburst uses the world name as the seed if it is not specified in the world config, requiring users to manually put a seed in every time a new world is loaded if they do not want to get the same world over and over again. This Pull Request cleans up some method names but most importantly adds the functionality of Vanilla-like random seed generation to replace the placeholder of using world names as seeds. (I'm not expecting this PR to be accepted, but rather to raise an issue in hopes that the Cloudburst maintainers might rectify this in the future, because this isn't a very complicated thing to implement but would be extremely frustrating for users if forgotten about when Cloudburst goes into production)

EDIT: If I may ask, what is planned for the default world loading? Does Cloudburst plan to have the default world options in Cloudburst yml? I was initially going to change the default and nether worlds to not depend on the yml file alongside the changes for seed generation but didn't do so in the end

Commit:
Changed World Generation to use random seed numbers if not specified in Config

TheShermanTanker committed:
Update Server.java

TheShermanTanker committed:
Squash whitespace

TheShermanTanker committed:
Add description as to why test was disabled

TheShermanTanker committed:
Squash another whitespace because I'm a f*cking dumbass

DaMatrix commented on src/main/java/org/cloudburstmc/server/Server.java:
```diff
@@ -222,6 +222,8 @@
 
     private static final Pattern UUID_PATTERN = Pattern.compile("^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}.dat$", Pattern.CASE_INSENSITIVE);
 
+    public static final Random RANDOM = new Random();
```
use `ThreadLocalRandom`

TheShermanTanker replied:
Changed to use ThreadLocalRandom

DaMatrix commented on src/main/java/org/cloudburstmc/server/Server.java:
```diff
@@ -1705,7 +1707,7 @@ private void registerVanillaComponents() {
     }
 
-    private void loadLevels() throws IOException {
+    private void loadWorlds() throws IOException {
```
this method is loading `Level` objects, not worlds.

TheShermanTanker replied:
Got it, will fix both issues

TheShermanTanker replied:
If I may ask though, is there a difference between Level objects and Worlds?

TheShermanTanker committed:
Change name of loadLevels method back to original

TheShermanTanker committed:
Changed to use ThreadLocalRandom

TheShermanTanker committed:
Update Server.java

TheShermanTanker committed:
Temporary commit please hold

TheShermanTanker committed:
Fix indentation error caused by github bug

TheShermanTanker commented:
Above issues rectified

TheShermanTanker committed:
Change potentially confusing error message

TheShermanTanker commented:
Bump

EDIT: Nevermind my dumbass didn't realize github doesn't work this way

DaMatrix commented on src/main/java/org/cloudburstmc/server/Server.java:
```diff
@@ -1717,41 +1718,43 @@ private void loadLevels() throws IOException {
             Object seedObj = config.getSeed();
             long seed;
+            if(seedObj == null) { //Auto generate seed if one was not specified
+                seedObj = ThreadLocalRandom.current().nextLong();
```
i haven't actually looked at this code in-depth for a while, but won't this be run every time the server starts, thus causing the seed to be different between restarts? the seed isn't written to config anywhere, and i'm pretty sure the seed stored in `level.dat` isn't used...

TheShermanTanker replied:
Whoops, looks like I forgot about that, thanks for pointing out the oversight. I would go for level.dat personally, but does Cloudburst prefer config instead?

TheShermanTanker commented:
@DaMatrix I've gotten back from checking the source and it does indeed seem that the seed stored inside level.dat is used if present- In fact, the seed inside level.dat will actually overwrite the seed from config and/or the randomly generated seed if the server can find it:
https://github.com/CloudburstMC/Server/blob/b000e760aebf294af426c616a8a53853a4708add/src/main/java/org/cloudburstmc/server/level/provider/leveldb/serializer/LevelDBDataSerializer.java#L145

DaMatrix commented:
alright, in that case i see no problems

SupremeMortal commented on src/main/java/org/cloudburstmc/server/Server.java:
```diff
@@ -1717,41 +1718,43 @@ private void loadLevels() throws IOException {
             throw new IllegalStateException("No worlds configured! Add a world to cloudburst.yml and try again!");
         }
-        List<CompletableFuture<Level>> levelFutures = new ArrayList<>(worldConfigs.size());
+        List<CompletableFuture<Level>> worldFutures = new ArrayList<>(worldConfigs.size());
```
Why was the name changed? Internally we stick to calling them levels since that's what vanilla calls them.

SupremeMortal commented on src/main/java/org/cloudburstmc/server/Server.java:
```diff
@@ -1717,41 +1718,43 @@ private void loadLevels() throws IOException {
+        CompletableFutures.allAsList(worldFutures).join();
 
-        //set default level
+        //Set default world
```
Same with these comment changes

TheShermanTanker commented:
@SupremeMortal I felt like worlds was a better name than "levels", as it is slightly more descriptive, but I'll change it back if required

TheShermanTanker committed:
Change names back

TheShermanTanker commented:
Changed the names back, I do still prefer the ring "worlds" has as opposed to "levels" though, I changed them because the config class is named World and the variable is worldConfigs, and I was confused why they're called 2 different things

TheShermanTanker commented:
Hi, just to check will the conflicts be automatically resolved on merge?

TheShermanTanker commented:
Computer crashed and took the entire repo with it, starting new commit :/
